### PR TITLE
[Relay] Add ADTs to text format

### PR DESCRIPTION
**Pull request by weberlo**
This PR tackles some of item 0 from [this issue](https://github.com/dmlc/tvm/issues/3584), using [this RFC](https://github.com/dmlc/tvm/issues/3016) as a reference.

One notable deviation is that `match` expressions can be specified as incomplete (i.e., not requiring all cases to be covered) by writing `match?`.

The tests in `test_ir_parser.py` currently do not pass if I don't inline every expression in the pretty printer.  This is because when graph vars are created, scoping becomes more complicated.  For example, the following program
```
def @length[A](%xs: List[A]) -> int32 {
    match? (%xs) {
    | Cons(_, %rest) => 1 + @length(%rest)
    | Nil => 0
    }
}
```
would be pretty-printed like so
```
def @length[A](%xs: List[A]) -> int32 {
    %0 = 1 + @length(%rest)
    match? (%xs) {
    | Cons(_, %rest) => %0
    | Nil => 0
    }
}
```
.  While this can still be parsed (by delaying evaluation of the graph expr until the context of its usage is available), it seems to add additional complexity without any benefits that I'm aware of.  It also makes testing equality with the original program more difficult.

Hopefully, this brings us closer, if not all the way there, to writing Relay's prelude in the text format.

CC @jroesch @MarisaKirisame @joshpoll @slyubomirsky @junrushao1994

### CODE REVIEW

File: python/tvm/relay/grammar/Relay.g4 [hunk 1]
```diff
@@ -17,7 +17,11 @@
  * under the License.
  */
 
-// list = *, seq = ?
+// TODO(weberlo): We need some way of indicating to users that you need to enable
```
Comment (slyubomirsky) [inline python/tvm/relay/grammar/Relay.g4]: Is a comment in this file (and one in the Relay docs) reminding users that the parser will not be updated unless the flag is set not sufficient?

File: python/tvm/relay/grammar/Relay.g4 [hunk 2]
```diff
@@ -74,106 +73,121 @@ METADATA: 'METADATA:' .*;
...
+defn
+  : 'def' globalVar typeParamList? '(' argList ')' ('->' typeExpr)? body  # funcDefn
+  | 'type' generalIdent typeParamList? '=' adtConsDefn+                      # adtDefn
+  ;
+
+constructorName: CNAME ;
+
+adtConsDefn: '|' constructorName ('(' typeExpr (',' typeExpr)* ')')? ;
+matchClause: '|' constructorName patternList? '=>' expr ;
+matchType : 'match' | 'match?' ;
```
Comment (slyubomirsky) [inline python/tvm/relay/grammar/Relay.g4]: I think you should clearly document the difference between these matches (complete vs incomplete) and ensure it round-trips through the pretty-printer

Comment (slyubomirsky) [reply python/tvm/relay/grammar/Relay.g4]: Ah, I notice the pretty-printer does handle it :+1:

File: python/tvm/relay/ty.py [hunk 3]
```diff
@@ -107,6 +107,7 @@ class Kind(IntEnum):
     Shape = 3
     Constraint = 4
     AdtHandle = 5
+    # TODO: This is never used.  Remove it?
```
Comment (slyubomirsky) [inline python/tvm/relay/ty.py]: I think we should keep it because if we need to implement more complicated kind-checking later, we could ensure nobody is sticking `TypeData` where it shouldn't be (anywhere but the module)

**Comment by slyubomirsky** [comment]
Thank you for adding tests for the additional features.

Higher-level point of discussion: What syntax do we want to settle on for text format type definitions? I like the style @weberlo used here (similar to OCaml's), but it may contrast a bit with other syntax we have (e.g., it doesn't use curly braces). 

Also on the subject of syntax, I think it may be a little redundant to include the line separators `|` between match clauses when we use braces to delimit a match block and each pattern must match exactly one expression. Do we want braces around the expression in the match clauses? As in, `pattern => { expr }`? Since there can only be one expression, the braces might be redundant, though.

One reason I am raising these questions is so we can update our documentation to be consistent with the syntax we actually implement.

File: python/tvm/relay/grammar/Relay.g4 (hunk 1 above)
Comment (weberlo) [reply python/tvm/relay/grammar/Relay.g4]: Yeah.  It probably is sufficient.  I was thinking more along the lines of a programmatic enforcement, but the number of people working on this area of the codebase doesn't really warrant that.

**Comment by weberlo** [comment]
@slyubomirsky I agree on the vertical bars in `match` exprs.  I was originally just implementing the syntax from the RFC and hoping to reopen the discussion in this PR.

If we remove the bars from `match`, then it falls more in line with Rust syntax.  We could then use Rustish syntax for ADT definitions.  An example Rust enum is
```rust
enum WebEvent {
    // An `enum` may either be `unit-like`,
    PageLoad,
    PageUnload,
    // like tuple structs,
    KeyPress(char),
    Paste(String),
}
```
and we could just replace `enum` with `type`.

We could also allow optional braces in match arms, which would be useful for sequencing.
```
match %ayy_lmao {
  Ayy => 0
  Lmao => {
    @mutate();;
    1
  }
}
```

The last minor point is that if we're going with Rust syntax, should we also require commas after match arms?

File: src/relay/ir/alpha_equal.cc [hunk 4]
```diff
@@ -215,7 +218,7 @@ class AlphaEqualHandler:
   bool VisitType_(const TypeVarNode* lhs, const Type& other) final {
     if (const TypeVarNode* rhs = other.as<TypeVarNode>()) {
       if (lhs->kind != rhs->kind) return false;
-      return LeafNodeEqual(GetRef<NodeRef>(lhs), other);
+      return lhs->var->name_hint == rhs->var->name_hint;
```
Comment (MarisaKirisame) [inline src/relay/ir/alpha_equal.cc]: typevar is equality by reference.

File: src/relay/ir/alpha_equal.cc [hunk 5]
```diff
@@ -288,7 +291,11 @@ class AlphaEqualHandler:
   }
 
   bool VisitType_(const GlobalTypeVarNode* lhs, const Type& other) final {
-    return GetRef<Type>(lhs) == other;
+    if (const GlobalTypeVarNode* rhs = other.as<GlobalTypeVarNode>()) {
+      // use name equality for global var for now.
+      return lhs->var->name_hint == rhs->var->name_hint;
```
Comment (MarisaKirisame) [inline src/relay/ir/alpha_equal.cc]: why this change?

**Comment by MarisaKirisame** [comment]
also:maybe change the name of graph_equal to something better (freevars_eq e.x.)

File: src/relay/ir/alpha_equal.cc (hunk 5 above)
Comment (weberlo) [reply src/relay/ir/alpha_equal.cc]: woops. both of these changes were just to get the tests passing. I didn't realize I should've been using `graph_equal`.

**Comment by weberlo** [comment]
@MarisaKirisame Is there not a more standard PL name for the type of equality that `graph_equal` encodes?

**Comment by MarisaKirisame** [comment]
@weberlo alpha_equal_up_to_free_vars.

File: python/tvm/relay/grammar/Relay.g4 [hunk 6]
```diff
@@ -17,7 +17,11 @@
...
  */
 
-// list = *, seq = ?
+/*
+ * NOTE: The `USE_ANTLR` option in `config.cmake` must be enabled in order for
+ * changes in this file to be reflected by the parser.
+ * NOTE: All upper-case rules are *lexer* rules and all camel-case rules are *parser* rules.
+ */
 
 grammar Relay;
```
Comment (MarisaKirisame) [inline python/tvm/relay/grammar/Relay.g4]: Update the version number to 0.0.4.

**Comment by slyubomirsky** [comment]
On syntax, I like the idea of optional braces but we should be certain that this will not cause problems for round-tripping. I think commas in type definitions and match definitions in the Rust style might be good in that they more closely resemble C-like syntax

**Comment by weberlo** [comment]
@slyubomirsky @MarisaKirisame I took a crack at implementing Rust-style syntax in the most recent commits.  Take a look when you get chance.

File: src/relay/ir/pretty_printer.cc [hunk 7]
```diff
@@ -778,8 +778,13 @@ class PrettyPrinter :
     }
     Doc separator;
     separator << "," << PrintNewLine();
-    doc << Brace(PrintSep(constructor_docs, separator) << ",");
-
+    Doc adt_body;
+    adt_body << PrintSep(constructor_docs, separator);
+    // add trailing comma if there are any constructors
```
Comment (MarisaKirisame) [inline src/relay/ir/pretty_printer.cc]: why trailing comma?

**Comment by weberlo** [comment]
I've just added support for external type declarations (example syntax is `extern Type[A]`), per @jroesch's suggestion.  It will allow for importing opaque types which can be interacted with via operators.

File: src/relay/ir/pretty_printer.cc (hunk 7 above)
Comment (weberlo) [reply src/relay/ir/pretty_printer.cc]: idk. it's not necessary. just a matter of style.

File: src/relay/ir/pretty_printer.cc [hunk 8]
```diff
@@ -343,13 +347,18 @@ class PrettyPrinter :
...
-             dg_.expr_node.at(expr)->parents.head->next);
+    auto it = dg_.expr_node.find(expr);
+    if (it == dg_.expr_node.end()) {
+      return true;
+    } else {
+      return !(it->second->parents.head && it->second->parents.head->next);
+    }
   }
 
+  // TODO(weberlo): Consolidate this method and `IsAtomic` in `pass_util.h`?
```
Comment (slyubomirsky) [inline src/relay/ir/pretty_printer.cc]: Might be a good idea if we can avoid repeating ourselves, though are we certain that these definitions (something we will always inline and atomic expressions) will always coincide?

File: tests/python/relay/test_ir_parser.py [hunk 9]
```diff
@@ -632,6 +641,236 @@ def test_tuple_type():
...
+    parse_text(
+        """
+        def @id[A](%x: A) -> A { x }
+        def @id[A](%x: A) -> A { x }
+        """
+    )
+
+
+def test_extern_adt_defn():
+    # TODO(weberlo): update this test once extern is implemented
```
Comment (slyubomirsky) [inline tests/python/relay/test_ir_parser.py]: Maybe we shouldn't include the syntax while we don't have the feature implemented (e.g., leave it for a different PR)

Comment (weberlo) [reply tests/python/relay/test_ir_parser.py]: @jroesch is planning on updating it very soon for the VM.  He just wanted me to add all of the necessary boilerplate.

Comment (slyubomirsky) [reply tests/python/relay/test_ir_parser.py]: okay

File: src/relay/ir/pretty_printer.cc (hunk 8 above)
Comment (weberlo) [reply src/relay/ir/pretty_printer.cc]: Yeah.  I was originally thinking it might be a good idea, but I'm not convinced they should be merged anymore.

File: tests/python/relay/test_ir_parser.py (hunk 9 above)
Comment (jroesch) [reply tests/python/relay/test_ir_parser.py]: I need this change ASAP, I just want the parsing changes to be shipped in a unit to keep things clean.
